File: elixir/leap/leap.exs
defmodule GuardSafe do
  defmacro divisible_by?(number, divisor) do
    quote do
      rem(unquote(number), unquote(divisor)) == 0
    end
  end
end

defmodule Year do
  import GuardSafe, only: [divisible_by?: 2]

  @doc """
  Returns whether 'year' is a leap year.

  A leap year occurs:

  on every year that is evenly divisible by 4
    except every year that is evenly divisible by 100
      except every year that is evenly divisible by 400.
  """
  @spec leap_year?(integer) :: boolean
  def leap_year?(year) when year |> divisible_by?(400), do: true
  def leap_year?(year) when year |> divisible_by?(100), do: false
  def leap_year?(year), do: year |> divisible_by?(4)
end

File: python/leap/leap.py
def leap_year(year):
    return basic_rule(year) and any(
        special_rule(year) for special_rule in [not_centennial, quadcentennial]
    )


def basic_rule(year):
    return divisible(year, 4)


def not_centennial(year):
    return not divisible(year, 100)


def quadcentennial(year):
    return divisible(year, 400)


def divisible(year, divisor):
    return year % divisor == 0
